apps/autoscheduler/LoopNest.cpp

```diff
@@ -0,0 +1,1861 @@
+                std::min(inlined_feat.innermost_pure_loop_extent,
+                         feat.innermost_pure_loop_extent);
+        } else {
+            inlined_feat.innermost_pure_loop_extent = feat.innermost_pure_loop_extent;
+        }
+        inlined_feat.inner_parallelism = 1;
+        inlined_feat.outer_parallelism = parallelism;
+    }
+}
+
+// Get the region required of a Func at this site, from which we
+// know what region would be computed if it were scheduled here,
+// and what its loop nest would be.
+const Bound &LoopNest::get_bounds(const FunctionDAG::Node *f) const {
+    if (bounds.contains(f)) {
+        const Bound &b = bounds.get(f);
+        // Expensive validation for debugging
+        // b->validate();
+        return b;
+    }
+    auto bound = f->make_bound();
+
+    // Compute the region required
+    if (f->is_output && is_root()) {
+        internal_assert(f->outgoing_edges.empty()) << "Outputs that access other outputs not yet supported\n";
+        // It's an output. Use the bounds estimate.
+        for (int i = 0; i < f->dimensions; i++) {
+            bound->region_required(i) = f->estimated_region_required[i];
+        }
+    } else {
+        internal_assert(!f->outgoing_edges.empty())
+            << "No consumers of " << f->func.name()
+            << " at loop over " << (is_root() ? "root" : node->func.name()) << '\n';
+        auto init = Span::empty_span();
+        for (int i = 0; i < f->dimensions; i++) {
+            bound->region_required(i) = init;
+        }
+
+        for (const auto *e : f->outgoing_edges) {
+            // Ignore consumers outside of this loop nest
+            if (!is_root() &&
+                (stage != e->consumer) &&
+                !stage->downstream_of(*(e->consumer->node))) {
+                continue;
+            }
+            const auto &c_bounds = get_bounds(e->consumer->node);
+
+            // Get the concrete sizes of the consuming loop
+            const auto *consumer_loop = &(c_bounds->loops(e->consumer->index, 0));
+
+            // Use the bounds relationship between the nodes to
+            // map from the consumer's loop to the required region
+            // of the producer.
+            e->expand_footprint(consumer_loop, &(bound->region_required(0)));
+        }
+    }
+
+    // Given a required region of this producer, use the bounds
+    // analysis to figure out what region actually gets
+    // computed. For most funcs, these are the same. Some things,
+    // like histograms or scans, you can only really compute all
+    // of at once.
+    f->required_to_computed(&(bound->region_required(0)), &(bound->region_computed(0)));
+
+    // Finally, figure out what loop nests will be used to compute
+    // this region.
+    for (int i = 0; i < (int)f->stages.size(); i++) {
+        f->loop_nest_for_region(i, &(bound->region_computed(0)), &(bound->loops(i, 0)));
+    }
+
+    const Bound &b = set_bounds(f, bound);
+    // b->validate();
```

Comment:
remove or justify keeping

Reply:
I added a comment about the cost of the calls to validate, as done at the beginning of the function.,